src/validation-engine.ts

```diff
@@ -0,0 +1,228 @@
+import {
+  ValidationResult,
+  FormValidationResult,
+  FieldsValidationSchema,
+  RecordValidationFunction,
+  createDefaultValidationResult,
+  RecordValidationFunctionSyncAsync,
+  isSyncValidationResult,
+  FieldValidationFunctionSyncAsync,
+  FieldValidationFunction,
+} from './model';
+
+import { isUndefinedOrNull } from './helper';
+
+import {
+  fireAllFieldsValidations,
+  fireSingleFieldValidations,
+  fireRecordValidations,
+} from './validation-dispatcher';
+import { buildFormValidationResult } from './form-validation-summary-builder';
+
+export class ValidationEngine {
+  private validationsPerField: FieldsValidationSchema = {};
+  private validationsGlobalForm: RecordValidationFunction[] = [];
+  private asyncValidationInProgressCount = 0;
+
+  // TODO, in new api we are adding an optional parameter
+  // to customize the error message (string array)
+  //
+  // TODO: Chain pattern applied here, check why it is useful
+  addFieldValidation(
+    key: string,
+    validation: FieldValidationFunctionSyncAsync,
+    customParams: any = {},
+    errorMessage?: string
+  ): ValidationEngine {
+    const asyncValidationFn = this.convertFieldValidationToAsyncIfNeeded(
+      validation
+    );
+
+    if (!this.isFieldKeyMappingDefined(key)) {
+      this.validationsPerField[key] = [];
+    }
+
+    this.validationsPerField[key].push({
+      validator: asyncValidationFn,
+      customArgs: customParams,
+      errorMessage: errorMessage,
+    });
+    return this;
+  }
+
+  // TODO: Should it be moved to model?
+  convertFieldValidationToAsyncIfNeeded(
+    validation: FieldValidationFunctionSyncAsync
+  ): FieldValidationFunction {
+    // Sugar we admit both flavors syncrhonous and asynchronous validators
+    return (values: any): Promise<ValidationResult> => {
+      const result = validation(values);
+
+      if (isSyncValidationResult(result)) {
+        return Promise.resolve(result as ValidationResult);
+      } else {
+        return result as Promise<ValidationResult>;
+      }
+    };
+  }
+
+  addRecordValidation(validation: RecordValidationFunctionSyncAsync): void {
+    // Sugar we admit both flavors syncrhonous and asynchronous validators
+    const validationAsync = this.convertFormValidationToAsyncIfNeeded(
+      validation
+    );
+
+    this.validationsGlobalForm.push(validationAsync);
+  }
+
+  // TODO: Should this be moved to model ?
+  convertFormValidationToAsyncIfNeeded(
+    validation: RecordValidationFunctionSyncAsync
+  ): RecordValidationFunction {
+    return (values: any): Promise<ValidationResult> => {
+      const result = validation(values);
+
+      if (isSyncValidationResult(result)) {
+        return Promise.resolve(result as ValidationResult);
+      } else {
+        return result as Promise<ValidationResult>;
+      }
+    };
+  }
+
+  public isValidationInProgress() {
```

Comment:
Do we keep this functionality?

I had never use it before, did you?

Reply:
I think it can be useful, altough Final Form loffers that